### ISSUE TITLE
Using Subresource Integrity with CDN URLs

### ISSUE THREAD

**Issue by jakirkham** [bokeh/bokeh#8397]
Would be nice for all CDN URLs provided by Bokeh in HTML documents to use [Subresource Integrity (SRI)]( https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity ). At this point basically [all browsers support SRI]( https://caniuse.com/#feat=subresource-integrity ) with the exception of IE (though Edge supports it). This would ensure that browsers checksum content from these CDN URLs and only use content if the checksums match.

**Comment by jakirkham** [comment]
Basically there are two tasks here as I see it.

1. Generate hashes for all files used from CDNs (probably at build time) and store them.
2. Inject these hashes into generated HTML documents.

**Comment by bryevdv** [comment]
> with the exception of IE 

By this do you mean that they are harmlessly ignored by IE? Or that adding them would cause Bokeh plots on IE to cease to function?

> Generate hashes for all files used from CDNs (probably at build time) and store them.

Store them where, exactly? I assume you mean the hashes need to be bundled with the package somehow, but want to be sure. 

> Inject these hashes into generated HTML documents.

I think there are more questions to answer. How does all this intersect with dev/rc builds, or with "setup.py develop" installs. I guess my preference would be that this only gets used for full releases but since it seems like things would be overly complicated otherwise. 

> So maybe some tweaking to how things are setup on the PyData CDN will be needed as well.

cc @aterrel do you have any ideas here? I don't currently have any visibility into the CDN

**Comment by jakirkham** [comment]
I'm not sure how IE would handle them if they showed up. Only putting into words what the chart shows me. If there is an IE user that could test this, that would be really informative.

Yes, by store I mean write them to files included in the package.

Not sure I follow. Could you please elaborate more on your concerns w.r.t. dev/rc builds?

**Comment by jakirkham** [comment]
Thanks Bryan.

If you are someone you know has access to IE, it should be straightforward for them to try that example linked from my previous comment to see if that works. Would be curious to learn the result.

**Comment by bryevdv** [comment]
> If you are someone you know has access to IE, it should be straightforward for them to try that example linked from my previous comment to see if that works. Would be curious to learn the result.

I don't, that's why I keep mentioning it.

**Comment by jakirkham** [comment]
Hmm...would have thought that was part of your release testing. How do you make sure releases work on IE now?

**Comment by bryevdv** [comment]
> Hmm...would have thought that was part of your release testing. How do you make sure releases work on IE now?

You are vastly overestimating the time and resources we have available. For micro releases we just make sure the appveyor test pass. Only for larger or more risky releases do we do manual windows testing. I have appealed now on multiple occasions asking for someone from the community to take on the role of Windows Maintainer but so far no one has decided to step up.

**Comment by martey** [comment]
Internet Explorer shouldn't be an issue with SRI. Browsers that don't support SRI still load the relevant scripts - see https://www.troyhunt.com/protecting-your-embedded-content-with-subresource-integrity-sri/:

> Thirdly and finally, if the browser doesn't support SRI... nothing happens. I mean nothing *different* happens in that the browser still loads the script, it just simply ignores the integrity tag.

**Comment by bryevdv** [comment]
@martey thanks for the information, I think given that knowledge, we can at least plan to proceed. 

I don't have access to the PyData CDN though, so some coordination will be involved. @aterrel have you accepted the invite that was sent out to add you as an admin?

@jakirkham just to be clear about this plan:

>Basically there are two tasks here as I see it.
> 
> Generate hashes for all files used from CDNs (probably at build time) and store them.
> Inject these hashes into generated HTML documents. 

you mean store them in the distributed packages that are published to pypi and conda?

**Comment by bryevdv** [comment]
@martey are you NF martey? I didn't not know :)

**Comment by bryevdv** [comment]
tentatively moving to `short term` but there is some infrastructure work that has to be done first.

**Comment by jakirkham** [comment]
Thanks @martey!

>> Generate hashes for all files used from CDNs (probably at build time) and store them.
>
> you mean store them in the distributed packages that are published to pypi and conda?

That's right, @bryevdv. Admittedly Conda will probably piggy-back on whatever is done for PyPI. Does that seem reasonable to you?

**Comment by bryevdv** [comment]
> Does that seem reasonable to you? 

I suppose? I have to be honest and state that I am in the middle of switching jobs and some other life changes at the moment. I'm probably not going to be able to personally do any work on this any time soon. I am happy to help anyone who does want to work on it, give pointers into our codebase, review a PR, help facilitate any related infrastructure work, etc.

**Comment by bryevdv** [comment]
@jakirkham @martey OK an update and a question. I am in the process of setting up *cdn.bokeh.org* to replace *cdn.pydata.org* (we will have to make sure the old links work basically forever, though) I am planning to use an s3 bucket and CloudFront. I have already set this up, in fact, e.g 

https://cdn.bokeh.org/bokeh/release/bokeh-1.2.0.js

Looking at the kinds of metadata that can be added to S3 obejcts, it does not appear that `Content-Security-Policy` is one of the available keys. At least, the AWS Console UI does not offer it in the list of metadata keys. Maybe a boto upload can set arbitrary headers? Do either of you know?

Alternatively, if CSP cannot be set on the server side, would it still be useful to the `integrity` attribute in `script` tags? (i.e. users would still benefit from digest comparision, even if the server side was not configured with CSP?)

Comments welcome, this is not an area of expertise for me.

**Comment by martey** [comment]
My experience with CloudFront is limited, but the references I could find online all suggest [using Lambda@Edge to add custom headers like `Content-Security-Policy`](https://aws.amazon.com/blogs/networking-and-content-delivery/adding-http-security-headers-using-lambdaedge-and-amazon-cloudfront/). You could also use a [meta tag](https://developers.google.com/web/fundamentals/security/csp/#the_meta_tag).

> Alternatively, if CSP cannot be set on the server side, would it still be useful to the integrity attribute in script tags? (i.e. users would still benefit from digest comparision, even if the server side was not configured with CSP?)

As far as I know, CSP is independent of SRI, so the latter should still work on websites not using the former.

---

This question might betray my complete lack of knowledge of how Bokeh and its CDN work, but *do you need to set a content security policy*?  I think that as long as the CDN supports [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS), using JavaScript snippets with SRI hashes should work.

This obviously wouldn't solve the problem of having Bokeh automatically add in SRI hashes when it creates JavaScript tags, but it would help people who manually add in hashes avoid the CORS error mentioned in https://github.com/bokeh/bokeh/issues/8397#issuecomment-436118730.

**Comment by bryevdv** [comment]
> As far as I know, CSP is independent of SRI, so the latter should still work on websites not using the former.

My understanding is that setting CSP means that the CDN would refuse to load except from script tags that provide an SRI to check. Which seemed like a good idea at first but now that I have though about it, does not seem like such a good idea. I am working nearly in the dark here (which is actually a large part of my reticence to act @jakirkham)  I'm working on updating our release automation to send files to the CloudFront CDN in parallel to the current CDN. Once that is done I can look into generating the hashes and baking them in to a build...

**Comment by bryevdv** [comment]
Given the recent sale of the `.org` name registry to a private equity firm, I have decided to prioritize this, in case our control of the `cdn.bokeh.org` domain is ever put in jeopardy. Here is my immedite plan, I would appreciate feedback @martey @jakirkham  @mattpap 

* [ ]  Generate hashes for all existing full release assets, using instructions https://www.srihash.org/*
* [ ] Create an `sri.json` file at the top level of the package that maps *[version][filename]* to a hash:
    ```
    {
        "1.4.0": [
            "bokeh.js": "<hash>",
            "bokeh.min.js": "<hash>"
            ...
        ],
        ...
    }
    ```
* [ ] Create a sphinx plugin to publish this information in a tabular form on a dedicated page in the docs
* [ ] Update the codepaths that generate `<script>` tags to look up and include an SRI hash for full release versions (raise an error if not? make this configurable?)

Here is the tricky part: where does the automation update and check in this file? The ideal place would be in `deploy.py` because that's less risky to test/develop and offers a change to abort a release if there are ever issues. But then we have to generate hashes on someone's *local* BokehJS, not on what the CI actually builds and uploads. Doing it in CI means pushing a commit from CI, lot's of chances for things to go wrong there. Also that commit would be *after* the release tag, as thing stand. 

I have another issue that proposes to split out building all the release assets from actually publishing/releasing. I think that will be a necessary first step to finishing this. Then the final "x.y.z" can be on a repo state that includes an updated `sri.json`.

**Comment by martey** [comment]
> But then we have to generate hashes on someone's local BokehJS, not on what the CI actually builds and uploads.

The files' content (and their hashes) shouldn't change depending on where they are generated, so you could generate hashes locally and use an automated test on the CI system to confirm that the hashes in `sri.json` are correct.

**Comment by mattpap** [comment]
This is actually a good opportunity to make sure that bundles are all the same on all platforms.

**Comment by bryevdv** [comment]
@jakirkham I have reconfigured `cdn.bokeh.org` to respect the necessary CORS policy. I was able to get your HTML file working with links to Bokeh 1.0.2 from the CDN (and it failed to work if I changed the version but not the hash). If you could also test things out to make sure there is not anything I've missed it would be very appreciated.  I should have a file SRI PR this weekend to close this issue but inserting the `integrity` attribute any time CDN resources are used.

**Comment by jakirkham** [comment]
That’s awesome! 🎉 Thanks Bryan! 😀 I’ll take a look and get back to you 🙂

**Comment by bryevdv** [comment]
@jakirkham t seems that some browsers may behave differently regarding CORS origin when viewing a local file. Chrome e.g. seems like it will refuse to load stating the origin is `null` Since we have no way of knowing if Bokeh content is being used in a page that is served, or viewed from a local file, and since viewing from a local file needs to to work, we may have no choice but to switch the new SRI integrity hash checks to be opt-in (e.g. env var to turn on). Do you have any other ideas?

**Comment by jakirkham** [comment]
Yeah that seems sensible.

**Comment by bryevdv** [comment]
well check out more details on #9773 I'm now not actually sure that might not make things worse :)

**Comment by jakirkham** [comment]
Ah ok. Will follow along there 🙂
